### DISCUSSION TITLE
packet format

### DISCUSSION THREAD

**Discussion by on1arf** [meshtastic/Discussions#71] [General]
Hi,
I am doing a writeup of meshtastic for a talk I plan in our local radioclub.

Can somebody explain the packet format of this packet?
```
DUMP: ff ff ff ff xx xx xx xx 7e 0f bd 95 e7 25 00 00 08 01 12 07 74 65 73 74 31 32 33 48 01 

ff ff ff ff -> destination (broadcast)
xx xx xx xx -> sender
7e 0f bd 95 -> packet id
e7 -> flags (hops)
25 -> channel hash
00 00 -> unused

08 ???
01 -> portnum ('01' = text message app)
12 ???
07 -> packet length
74 65 73 74 31 32 33 -> 'test123'
48 01 ???
```

What is the meaning of the 4 bytes marked with a question-mark?

I did read the protobufs source-code and documentatiion and the source-code of the firmware but sofar I have not managed to actualy "decode" it.

**Comment by ianmcorvidae** [comment]
It's important to understand that protocol buffers are a "tag-length-value" encoding scheme (with some clever tricks that allow omitting the length in many cases). The protobuf-encoded part of your dump starts at the `08`, which is the tag indicating the following data corresponds to the portnum field. For that, the first five bytes are the field number (1) and the last three are the "wire type" (in this case, 0, for varint encoding (variable length integer)). 12 (field 2, with length encoding) and 48 (field 9, varint) are also tags, and the final 01 is the data for field 9 (which is the bitfield in which some things are encoded more compactly than in booleans in protobuf -- right now, the "ok to MQTT" flag is the thing that lives there and nothing else).

**Reply by ianmcorvidae** [reply]
I'm guessing you're using ham mode or otherwise unencrypted, also -- normally everything after the header is encrypted over the air and must be decrypted before protobuf decoding.

**Reply by on1arf** [reply]
yes, I am in hamradio mode.
(as this is a talk for a hamradio club, this is for "meshtastic from a hamradio perspective".

Thanks for the info.
Looks like I missed the 'where does this protobuf' fit into this?' part.

I'll go looking for the list that maps the protobuf tags to their meaning so get a better understand if the different fields

Thanks! (thumbs-up)
